src/ImageSharp/Formats/Qoi/QoiDecoderCore.cs

```diff
@@ -0,0 +1,287 @@
+        {
+            for (int j = 0; j < this.header.Width; j++)
+            {
+                byte operationByte = (byte)stream.ReadByte();
+                byte[] pixelBytes;
+                Rgba32 readPixel;
+                TPixel pixel = default;
+                switch ((QoiChunk)operationByte)
+                {
+                    // Reading one pixel with previous alpha intact
+                    case QoiChunk.QoiOpRgb:
+                        pixelBytes = new byte[3];
+                        if (stream.Read(pixelBytes) < 3)
+                        {
+                            ThrowInvalidImageContentException();
+                        }
+
+                        readPixel = previousPixel with { R = pixelBytes[0], G = pixelBytes[1], B = pixelBytes[2] };
+                        pixel.FromRgba32(readPixel);
+                        pixelArrayPosition = GetArrayPosition(readPixel);
+                        previouslySeenPixels[pixelArrayPosition] = readPixel;
+                        break;
+
+                    // Reading one pixel with new alpha
+                    case QoiChunk.QoiOpRgba:
+                        pixelBytes = new byte[4];
+                        if (stream.Read(pixelBytes) < 4)
+                        {
+                            ThrowInvalidImageContentException();
+                        }
+
+                        readPixel = new Rgba32(pixelBytes[0], pixelBytes[1], pixelBytes[2], pixelBytes[3]);
+                        pixel.FromRgba32(readPixel);
+                        pixelArrayPosition = GetArrayPosition(readPixel);
+                        previouslySeenPixels[pixelArrayPosition] = readPixel;
+                        break;
+
+                    default:
+                        switch ((QoiChunk)(operationByte & 0b11000000))
+                        {
+                            // Getting one pixel from previously seen pixels
+                            case QoiChunk.QoiOpIndex:
+                                readPixel = previouslySeenPixels[operationByte];
+                                pixel.FromRgba32(readPixel);
+                                break;
+
+                            // Get one pixel from the difference (-2..1) of the previous pixel
+                            case QoiChunk.QoiOpDiff:
+                                byte redDifference = (byte)((operationByte & 0b00110000) >> 4),
+                                    greenDifference = (byte)((operationByte & 0b00001100) >> 2),
+                                    blueDifference = (byte)(operationByte & 0b00000011);
+                                readPixel = previousPixel with
+                                {
+                                    R = (byte)((previousPixel.R + (redDifference - 2)) % 256),
+                                    G = (byte)((previousPixel.G + (greenDifference - 2)) % 256),
+                                    B = (byte)((previousPixel.B + (blueDifference - 2)) % 256)
```

Comment:
You can apply the modulo trick I commented later. `mod 256` should be equivalent to `& 15˙ if I'm not mistaken. Would be good to have a `Numerics.Modulo256` method.